common/tdx/verifier.c

```diff
@@ -0,0 +1,584 @@
+    result = OE_OK;
+
+done:
+    if (claims)
+        _free_claims_list(NULL, claims, claims_length);
+
+    return result;
+}
+
+// Gets the optional format settings for the given verifier plugin context.
+// For SGX local attestation, this would be the sgx_target_info_t struct.
+static oe_result_t _get_format_settings(
+    oe_verifier_t* context,
+    uint8_t** settings,
+    size_t* settings_size)
+{
+    OE_UNUSED(context);
+    OE_UNUSED(settings);
+    OE_UNUSED(settings_size);
+    return OE_UNSUPPORTED;
+}
+
+static oe_result_t _verify_report(
+    oe_verifier_t* context,
+    const uint8_t* report,
+    size_t report_size,
+    oe_report_t* parsed_report)
+{
+    OE_UNUSED(context);
+    OE_UNUSED(report);
+    OE_UNUSED(report_size);
+    OE_UNUSED(parsed_report);
+    return OE_UNSUPPORTED;
+}
+
+static oe_result_t _get_input_time(
+    const oe_policy_t* policies,
+    size_t policies_size,
+    oe_datetime_t** time)
+{
+    if (!policies)
+    {
+        *time = NULL;
+        return OE_OK;
+    }
+
+    for (size_t i = 0; i < policies_size; i++)
+    {
+        if (policies[i].type == OE_POLICY_ENDORSEMENTS_TIME)
+        {
+            if (policies[i].policy_size != sizeof(**time))
+                return OE_INVALID_PARAMETER;
+
+            *time = (oe_datetime_t*)policies[i].policy;
+            return OE_OK;
+        }
+    }
+
+    // Time not found, which is fine since it's an optional parameter.
+    *time = NULL;
+    return OE_OK;
+}
+
+static oe_result_t _verify_evidence(
+    oe_verifier_t* context,
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size,
+    const uint8_t* endorsements_buffer,
+    size_t endorsements_buffer_size,
+    const oe_policy_t* policies,
+    size_t policies_size,
+    oe_claim_t** claims,
+    size_t* claims_length)
+{
+    oe_result_t result_verify_quote = OE_UNEXPECTED;
+    const uint8_t* custom_claims_buffer = NULL;
+    size_t custom_claims_buffer_size = 0;
+    oe_result_t result = OE_UNEXPECTED;
+    oe_uuid_t* format_id = NULL;
+    oe_datetime_t* time = NULL;
+
+    if (!context || !evidence_buffer || !evidence_buffer_size ||
+        (!endorsements_buffer != !endorsements_buffer_size) ||
+        (!policies != !policies_size) || (!claims != !claims_length))
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    format_id = &context->base.format_id;
+
+    OE_CHECK(_get_input_time(policies, policies_size, &time));
+
+    if (!memcmp(format_id, &_uuid_tdx_quote_ecdsa, sizeof(oe_uuid_t)))
+    {
+        tdx_quote_t* quote = (tdx_quote_t*)evidence_buffer;
+
+        // TDX quote should have version 4
+        if (evidence_buffer_size < sizeof(*quote) + quote->signature_len ||
+            quote->version != SGX_QE4_QUOTE_VERSION ||
+            quote->sign_type != SGX_QL_ALG_ECDSA_P256 ||
+            quote->tee_type != TDX_QUOTE_TYPE)
+            OE_RAISE(OE_INVALID_PARAMETER);
+    }
+    else
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK_NO_TCB_LEVEL(
```

Comment:
No, we will not support endorsement for TDX verification.

Reply:
@mingweishih can I ask why not? we use that functionality in CCF to offload the cost of fetching the endorsements away from the primary and to the joining nodes.

Reply:
@achamayou Do you have the same requirement for TDX? The reason is that OE is planning to using QVL for both SGX and TDX for the long-term so that we can eliminate the cost of maintenance.  Endorsement handling is one area that OE has lots of duplicated code with QVL, so we want to get rid of that (for TDX only now).

Reply:
@mingweishih yes, the requirement applies regardless of the TEE we use, we do it for SEV SNP as well. Aside from the performance impact, we also persist the endorsements to the ledger to enable offline audit, and we would lose that if we switched to a verification method that doesn't expose them.

I understand and do sympathize with the desire to deduplicate code, is it perhaps possible to request that QVL supports passing in endorsements? We're very happy to switch formats/APIs, we don't mind that at all, but the functionality itself is important for us.

Reply:
For SGX, how does CCF fetch the endorsements today?

Reply:
I would assume you get the endorsement as part of `oe_get_evidence`.  Given we don't have get the evidence generation API for TDX, we would need to add a new API for TDX if needed.

Reply:
@mingweishih with `oe_get_evidence`: https://github.com/microsoft/CCF/blob/82ea63a5cd4da3ec7f1032645b6089b1a866d42f/include/ccf/pal/attestation.h#L347

Reply:
> I would assume you get the endorsement as part of `oe_get_evidence`. Given we don't have get the evidence generation API for TDX, we would need to add a new API for TDX to fetch endorsement if needed.

Is the intention that OE would only support verification but not getting evidence? If so, what is the recommended way to get TDX evidence?

Reply:
+ @jazzybluesea for question on the TDX quote/evidence generation.

Reply:
@achamayou /@mingweishih : We will have a document soon that will document the endpoint/library for getting the TDX quote/evidence.